LPS-127012, LPS-129571 Create BalloonEditor component and sample portlet

julien opened the pull request:
These are just the base files for our `BalloonEditor` component, which
contains a React component using CKEditor's `balloontoolbar` plugin with the
minimum required configuration, a Java class and its corresponding JSP.

Additionally this PR contains a sample portlet in the `frontend-editor-ckeditor-sample-web`
module: the purpose of this portlet, is to be able to test our changes
during the implementation of the `BalloonEditor` component.

To add this portlet, edit the main page (or either create another widget page)
and look for CKEditor Sample in the widgets section.

For the moment this is what it looks like:

![frontend-editor-ckeditor-sample-web](https://user-images.githubusercontent.com/5572/114551106-3d5afa80-9c63-11eb-8048-583787abb02f.png)

**Notes**
- This PR was originally sent and reviewed [here](https://github.com/liferay-frontend/liferay-portal/pull/932),
but got reverted [here](https://github.com/liferay-frontend/liferay-portal/pull/964) to fix [LPS-130399](https.//issues.liferay.com/browse/LPS-130399)

- For the moment I think this should be a draft PR to avoid it from being 
forwarded before everyone has had a change to have a look at it

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-17.liferay.com/job/test-portal-source-format/4951/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-17.liferay.com//userContent/jobs/test-portal-source-format/builds/4951/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8147/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-16.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/5dff0e4b8f856d2f4f83453ee62cd3081dcbd083">5dff0e4b8f856d2f4f83453ee62cd3081dcbd083</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-127012-LPS-129571">LPS-127012-LPS-129571</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/8a557da90ec1fb703c10666ea65c903399e2d1ff">8a557da90ec1fb703c10666ea65c903399e2d1ff</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-17.liferay.com/job/test-portal-source-format/4951/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-17.liferay.com/userContent/jobs/test-portal-source-format/builds/4951/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-17.liferay.com/job/test-portal-source-format/4951/">test-portal-source-format#4951</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-17.liferay.com/userContent/jobs/test-portal-source-format/builds/4951/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/968">liferay-frontend#968</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1525384745">[master] ci:test:sf - julien &gt; liferay-frontend - PR#968 - 2021-04-13[05:25:06]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-17.liferay.com/job/publish-testray-report/223/">publish-testray-report#223</a><br/></div></div>

edalgrin commented:
Just started reviewing :)

jbalsas commented on modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:
```diff
@@ -0,0 +1,116 @@
+
+				initialCssClass = CKEDITOR.env.cssClass;
+
+				CKEDITOR.env.cssClass = 'lfr_balloon_editor';
```
I'm wondering if this should be `CKEDITOR.env.cssClass += ' lfr_balloon_editor'` or `CKEDITOR.env.cssClass = `${CKEDITOR.env.cssClass} lfr_balloon_editor`.

Also I think our convention is to use dashes instead of underscores for our CSS classes. I'd consult an expert here, but we might want to make this `lfr-balloon-editor`. I know CKEditor's approach is different, but since this is our own code, I think it makes sense it follows our own rules.

julien replied:
About the `CKEDITOR.env.cssClass`, you're right: it might be dangerous to replace it with our own class ( @ambrinchaudhary asked me the same thing via Slack) 

I'll got with dashes in the class name, and use our conventions as much as possible.

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:stable - 9 out of 9 jobs passed</h3><h3>:heavy_check_mark: ci:test:relevant - 23 out of 23 jobs passed in 2 hours 33 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/5dff0e4b8f856d2f4f83453ee62cd3081dcbd083">5dff0e4b8f856d2f4f83453ee62cd3081dcbd083</a></p><details><summary><strong>ci:test:stable - 9 out of 9 jobs PASSED</strong></summary><details><summary><strong>9 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430870/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430872/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-21.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/410444/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439777/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439775/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430874/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-9.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/382409/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439774/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/466017/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><details><summary><strong>ci:test:relevant - 23 out of 23 jobs PASSED</strong></summary><details><summary><strong>23 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8147/">test-portal-acceptance-pullrequest(master)</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430873/">test-portal-acceptance-pullrequest-batch(master)/central-requirements-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439776/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk11_open/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430870/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/493756/">test-portal-acceptance-pullrequest-batch(master)/functional-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/466018/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430872/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-21.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/410444/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/469524/">test-portal-acceptance-pullrequest-batch(master)/lpkg-base-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/469523/">test-portal-acceptance-pullrequest-batch(master)/lpkg-controller-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/493755/">test-portal-acceptance-pullrequest-batch(master)/modules-compile-jdk8/0</a></li><li><a href="https://test-1-19.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459278/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439777/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-17.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459005/">test-portal-acceptance-pullrequest-batch(master)/modules-semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-17.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459006/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439775/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430871/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-project-templates-jdk8/0</a></li><li><a href="https://test-1-8.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/430874/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-9.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/382409/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-9.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/382410/">test-portal-acceptance-pullrequest-batch(master)/semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439774/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-7.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/439773/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8/0</a></li><li><a href="https://test-1-6.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/466017/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><h5>For more details click <a href="https://test-1-16.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/jenkins-report.html">here</a>.</h5></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-16/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-16.liferay.com/job/test-portal-acceptance-pullrequest(master)/8147/">test-portal-acceptance-pullrequest(master)#8147</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-16.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8147/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>relevant<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/968">liferay-frontend#968</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1525452104">[master] ci:test:relevant - julien &gt; liferay-frontend - PR#968 - 2021-04-13[05:25:05]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-16.liferay.com/job/publish-testray-report/239/">publish-testray-report#239</a><br/></div></div>

ambrinchaudhary commented:
Hey @liferay-frontend! I just marked this pull as ready for review, as I need it for rebasing another [pull](https://github.com/liferay-frontend/liferay-portal/pull/960). So any comment would be more than appreciated before forwarding it. Thanks all!

wincent commented on modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json:
```diff
@@ -0,0 +1,15 @@
+		"react": "16.12.0",
+		"react-dom": "16.12.0"
+	},
+	"name": "frontend-editor-ckeditor-sample-web",
```
This should be named `@liferay/frontend-ckeditor-sample-web` to guard against [dependency confusion attacks](https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/general/security/dependency_confusion.md).

We [would like to be able to lint for this](https://github.com/liferay/liferay-frontend-projects/issues/469), but we can't do that until we've renamed all of our existing packages ([LPS-129670](https://issues.liferay.com/browse/LPS-129670) (or we implement a whitelist). cc @jbalsas, @dsanz about prioritization of that work. (My feeling: this will be soul-destroying grunt work, but I'll feel nervous until we do it... and if we don't want to do it, we should hack in the whitelist so that we can deploy a lint soon.)

dsanz replied:
Having a white list may be read as "we want these to not have a namespace", when perhaps it should be taken as "that's the amount of pending work". Could we differentiate those in some way? And, are there chances that, once we rename all of our packages, whitelist becomes empty?

wincent replied:
> Could we differentiate those in some way? And, are there chances that, once we rename all of our packages, whitelist becomes empty?

Yeah, the "white list" (which I should probably call "allow list") means "it's ok for these to not be in a named scope" _and_ it will start off by coinciding with "the list of packages that are pending rename". And yes, once everything is renamed, the whitelist can be deleted. (I wouldn't actually be bothered with emptying it along the way... I'd just remove the whole thing at the end.)

julien replied:
@wincent sorry about that, I really need to remember this rule, I don't want you (or anyone else) to have to rename a gazillion packages again

wincent replied:
The worry isn't about "gazillion" packages. Say Liferay adds "a few" packages (5, 10?) over the next few months. All we need is _one_ [bad actor](https://en.wikipedia.org/wiki/Nicolas_Cage) to come along and squat the package name on npm and we've got a potential headache. The odds of it happening are slim, and the scope of the likely damage if it does happen is not that great, but that's cold comfort because the badness of the worst case scenario is bad enough. `#bad`

![bad](https://user-images.githubusercontent.com/7074/114839998-a911a480-9dd6-11eb-9f03-71a29032a02c.gif)

wincent commented on modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:
```diff
@@ -0,0 +1,116 @@
+		};
+	};
+
+	let initialCssClass;
```
Just to note a potential point of fragility here. _If_ this component ever re-rerenders, and _if_ it doesn't reliably trigger `onBeforeLoad` and `onDestroy` in perfect matching pairs, it is possible that the value that gets restored on line 51 will not be the one that got saved on line 46...

And also, that looks like a global variable... what's going to happen if there are multiple editors on the page fighting to save/restore the "initial" class? 😱 (Maybe this was discussed previously, so please forgive me if I missed that.)

julien replied:
> And also, that looks like a global variable...
Yes, once more thing I didn't think about correctly

> what's going to happen if there are multiple editors on the page fighting to save/restore the "initial" class?
Probably something I don't want :joy:

> (Maybe this was discussed previously, so please forgive me if I missed that.)

It hasn't been discussed so thanks for the heads up. 
I'm thinking this should be in the component's "state" (i.e. `useState`)

wincent commented on modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:
```diff
@@ -0,0 +1,116 @@
+					buttons:
+						'Bold,Italic,Underline,RemoveFormat,Link,NumberedList,BulletedList,JustifyLeft,JustifyCenter,JustifyRight,JustifyBlock,Anchor',
+					refresh(editor, path) {
+						const tags = new Set([
+							'h1',
+							'h2',
+							'h3',
+							'p',
+							'span',
+							'strong',
+							'li',
+						]);
```
Using a `Set`.

![god-bless-you-obama](https://user-images.githubusercontent.com/7074/114747926-a3be4680-9d51-11eb-8b02-a679b43c5926.gif)

wincent commented on modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:
```diff
@@ -0,0 +1,116 @@
+
+						let result = false;
+
+						tags.forEach((tag) => {
+							if (path.contains(tag)) {
+								result = true;
+							}
+						});
```
The `Set` is good for `O(1)` insertion, deletion, membership testing, but you are using it here for iteration `O(n)` (via `forEach`) and then you have a `path.contains()` which is itself a linear scan, so that's quadratic.

I want my Obama gif back!

What kind of object is `path`?

julien replied:
To explain my intent (and give you back your gif): I thought of using a `Set` because I want each element to be unique, and now I'm thinking that, since I'm "adding" those items myself and that nothing from "outside" is going to touch `tags`, I could probably use a regular `Array`.

When it comes to iteration, you're right, that's something I didn't really think about.

The `path` object is a [`CKEDITOR.dom.elementPath`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_elementPath.html)(that's a link to the docs, but I guess you also want to have a look at the [source](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/core/dom/elementpath.js)), and the idea behind this iteration is the following:
If the `path` contains any of those "tags", then we should show this toolbar ... 

I actually changed [this part of the code](https://github.com/liferay-frontend/liferay-portal/blob/12dac565f1b763bf2966e5d391dadf03c05d341a/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js#L58-L82) locally, to be like this

```js
balloonToolbars.create({
    buttons: 'Bold,Italic,Underline,RemoveFormat,Link,NumberedList,BulletedList,JustifyLeft,JustifyCenter,JustifyRight,JustifyBlock,Anchor',
    cssSelector: '*',
});
```

(*It's not something I came up with, since it was already "like" this in the original "draft"; and unfortunately, I can't find that pull request, and I can't remember where it was ... but I think it doesn't really matter*)


The good news is that the result is exactly the same.  This also looks more "simple" even though I haven't analyzed what it's doing "under the hood", but it's probably better for now than my quadratic monster.

*Another thing I didn't say (and probably should have because it's not obvious), is that this will probably change. I'm still trying to figure out how to use this CKEditor plugin to behave like we want it to.*  

![](https://media.giphy.com/media/3o7TKCRsacaH9u7HcA/source.gif)

wincent commented:
> So any comment would be more than appreciated before forwarding it. Thanks all!

Given that we had to revert this once, I think we should wait for the release to be 100% confirmed before we merged. It's just not worth the risk of getting it wrong a second time (because it would get us onto _la liste de merde_, as @julien would say)... 😂

julien commented:
I need to (fully) read and reply to your other comments, but I wanted to reply on this one first.

I totally agree that we could wait before merging this in right now, because

1. It's not finished 
2. @ambrinchaudhary has access to my liferay-portal fork so she can fetch, push or even delete the whole thing. 
3. I've skimmed through the comments @wincent left and there is stuff that needs to be fixed

I think there's no problem having this "sitting" here for a while and being a bit more patient: I definitely don't want to be on that list :joy:

ambrinchaudhary commented:
Ok, then I guess is better to put this again as draft, to avoid accidentally forwards. 

Just a note @julien, some comments like in https://github.com/liferay-frontend/liferay-portal/pull/968#discussion_r613413763 may not worth to fix because I changed them in https://github.com/liferay-frontend/liferay-portal/pull/960. Maybe we should merge this two pulls in just one...

ambrinchaudhary commented:
Just started reviewing :)

julien commented:
I've had another thought about this PR (the good thing about these draft PRs is that we can actually take time to think about them) and I'm thinking we probably don't want and don't need the sample (`frontend-editor-ckeditor-sample-web`) portlet  for the following reasons:

1. It adds more code to maintain (for little value)
2. We're also planning to replace the current editor in `blogs-web` and page editor, so we'll already have at least more than one place to test this new editor

What do you think about that @ambrinchaudhary?
If you agree, I'll update the PR and remove the corresponding commits.

ambrinchaudhary commented:
Hey @julien, 

for me this `frontend-editor-ckeditor-sample-web` portlet is very useful, it allows us to test this BalloonEditor properly in portal (you know it is not the same to be working on a demo that in portal). It does not add to much to maintain, because it only renders a taglib (`<liferay-editor:editor>`), and as we thought at the beginning, we can extend it to not just render the BalloonEditor, but any other editors available in portal so this could be also usefull for any developer or QAs. 

Also.... I don't think we can replace the Balloon Editor in Blogs or Page Editor until most of the work is done (which as you know won't be soon), so until then, if this portlet is not available there is no way to test the new editor in portal and face problems sooner.

jbalsas commented:
Hey @julien, @ambrinchaudhary, one option to solve this would be to treat this sample portlet similar to what @zhangyan0701 and @john-co did with https://github.com/liferay-frontend/liferay-portal/tree/master/modules/apps/frontend-taglib/frontend-taglib-clay-test-alert-toast-sample-web

If we considered this a QA test portlet, then they could take it over and use it to build future coverage for our WYSIWYG editors.

julien commented:
@jbalsas I actually didn't think about it but it probably make sense.
I'll check what needs to be changed to make that happen. @ambrinchaudhary, I guess you would also be happy about doing it this way, am I right?

julien commented:
@ambrinchaudhary I removed the `.lfr-buildportal` file as I didn't see it [in the module QA is using to test toasts](https://github.com/liferay-frontend/liferay-portal/tree/master/modules/apps/frontend-taglib/frontend-taglib-clay-test-alert-toast-sample-web):bread:. I also rebased with `master` just to make sure everything was up to date.

ambrinchaudhary commented:
LGTM @julien, thanks! I'm ok about sending this to master ;-)

wincent commented:
> I'm ok about sending this to master

![join-the-fun](https://user-images.githubusercontent.com/7074/116090507-3e008180-a6a4-11eb-9b4c-292244122995.gif)

`#AllTheCoolKidsAreDoingIt`

ambrinchaudhary commented:
As work is done here and release is completed, going to forward this pull 🎉

ambrinchaudhary commented:
ci:forward

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

The pull request will automatically be forwarded to the user `brianchandotcom` if the following test suites pass:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**relevant**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**stable**

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-14.liferay.com/job/test-portal-source-format/4876/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-14.liferay.com//userContent/jobs/test-portal-source-format/builds/4876/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 4 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/4b4ff53717645e7bf4c4b373b5cb5d218fb1207a">4b4ff53717645e7bf4c4b373b5cb5d218fb1207a</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/julien/liferay-portal/tree/LPS-127012-LPS-129571">LPS-127012-LPS-129571</a><br/>Branch GIT ID: <a href="https://github.com/julien/liferay-portal/commit/5faa55c364cd45ae8a2533da8de639885ad10544">5faa55c364cd45ae8a2533da8de639885ad10544</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-14.liferay.com/job/test-portal-source-format/4876/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-source-format/builds/4876/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-14.liferay.com/job/test-portal-source-format/4876/">test-portal-source-format#4876</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-source-format/builds/4876/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/968">liferay-frontend#968</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1528736141">[master] ci:test:sf - julien &gt; liferay-frontend - PR#968 - 2021-04-27[00:12:04]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-14.liferay.com/job/publish-testray-report/482/">publish-testray-report#482</a><br/></div></div>

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>relevant</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest(master)/8124/">test-portal-acceptance-pullrequest(master)</a></p><p><b>For more details click <a href="https://test-1-14.liferay.com//userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:stable - 9 out of 9 jobs passed</h3><h3>:heavy_check_mark: ci:test:relevant - 23 out of 23 jobs passed in 5 hours</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/4b4ff53717645e7bf4c4b373b5cb5d218fb1207a">4b4ff53717645e7bf4c4b373b5cb5d218fb1207a</a></p><h4>Upstream Comparison:</h4><p>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/4b4ff53717645e7bf4c4b373b5cb5d218fb1207a">4b4ff53717645e7bf4c4b373b5cb5d218fb1207a</a><br/>Jenkins Build URL: <a href="https://test-1-1.liferay.com/job/test-portal-acceptance-upstream-dxp(master)//1801">Acceptance Upstream DXP (master) #1801</a></p><details><summary><strong>ci:test:stable - 9 out of 9 jobs PASSED</strong></summary><details><summary><strong>9 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470899/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495292/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470904/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459940/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224654/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470905/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224656/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495288/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470901/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><details><summary><strong>ci:test:relevant - 23 out of 23 jobs PASSED</strong></summary><details><summary><strong>23 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest(master)/8124/">test-portal-acceptance-pullrequest(master)</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470903/">test-portal-acceptance-pullrequest-batch(master)/central-requirements-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495290/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk11_open/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470899/">test-portal-acceptance-pullrequest-batch(master)/functional-smoke-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224657/">test-portal-acceptance-pullrequest-batch(master)/functional-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495293/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495292/">test-portal-acceptance-pullrequest-batch(master)/functional-upgrade-tomcat90-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470904/">test-portal-acceptance-pullrequest-batch(master)/js-unit-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495294/">test-portal-acceptance-pullrequest-batch(master)/lpkg-base-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470902/">test-portal-acceptance-pullrequest-batch(master)/lpkg-controller-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495289/">test-portal-acceptance-pullrequest-batch(master)/modules-compile-jdk8/0</a></li><li><a href="https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459939/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8/0</a></li><li><a href="https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459940/">test-portal-acceptance-pullrequest-batch(master)/modules-integration-mysql57-jdk8_stable/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224655/">test-portal-acceptance-pullrequest-batch(master)/modules-semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495291/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224654/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-jdk8_stable/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470900/">test-portal-acceptance-pullrequest-batch(master)/modules-unit-project-templates-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470905/">test-portal-acceptance-pullrequest-batch(master)/pmd-jdk8/0</a></li><li><a href="https://test-1-18.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/224656/">test-portal-acceptance-pullrequest-batch(master)/rest-builder-jdk8/0</a></li><li><a href="https://test-1-5.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/459941/">test-portal-acceptance-pullrequest-batch(master)/semantic-versioning-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495288/">test-portal-acceptance-pullrequest-batch(master)/service-builder-jdk8/0</a></li><li><a href="https://test-1-3.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/495287/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8/0</a></li><li><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest-batch(master)/470901/">test-portal-acceptance-pullrequest-batch(master)/unit-jdk8_stable/0</a></li></ul></details></details><h5>For more details click <a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/jenkins-report.html">here</a>.</h5></details></html><h5>Test bundle downloads:</h5><ul><li><a href="https://files.liferay.com/private/bundles/test-1-14/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/liferay-portal-bundle-tomcat.tar.gz">liferay-portal-bundle-tomcat.tar.gz</a> (mirrors: <a href="http://mirrors.dlc.liferay.com/files.liferay.com/private/bundles/test-1-14/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/liferay-portal-bundle-tomcat.tar.gz">dlc</a>, <a href="http://mirrors.lax.liferay.com/files.liferay.com/private/bundles/test-1-14/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/liferay-portal-bundle-tomcat.tar.gz">lax</a>)</li></ul>

liferay-continuous-integration commented:
All required test suite(s) passed.
Forwarding pull request to `brianchandotcom`.
[Console](https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest(master)/8124//consoleFull)

liferay-continuous-integration commented:
Pull request has been successfully forwarded to https://github.com/brianchandotcom/liferay-portal/pull/101141
[Console](https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest(master)/8124//consoleFull)

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-14.liferay.com/job/test-portal-acceptance-pullrequest(master)/8124/">test-portal-acceptance-pullrequest(master)#8124</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-14.liferay.com/userContent/jobs/test-portal-acceptance-pullrequest(master)/builds/8124/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>relevant<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-frontend/liferay-portal/pull/968">liferay-frontend#968</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=1528778841">[master] ci:test:relevant - julien &gt; liferay-frontend - PR#968 - 2021-04-27[01:10:16]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-14.liferay.com/job/publish-testray-report/494/">publish-testray-report#494</a><br/></div></div>
